### PR TITLE
Model config client facades

### PR DESCRIPTION
**Pull request by tlm**
With this change we are now making the model config client facade use DQlite backed services for reading, writing and removing model config values.

All of the model config checking logic bar one item has been moved down into the services layer. The remaining one requires permission checks so we have left that as an API concern.

The storage and secret backend checkers need to be fully implemented as well. This will be worked on with Wallyworld in a follow up PR.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

### Test 1
The first test is we need to make sure that model config is the same between this change and the last commit.

1. Bootstrap a controller off of main and add a new model.
2. Run `juju model-config` and save the output.
3. Bootstrap a controller with this branch and add a new model.
4. Run `juju model-config`
5. Compare the output. They should be the same with the exception of storage. This will be dealt with in a follow up PR.

### Test 2
1. Bootstrap a controller and a new model.
2. Set model config values for the new model such as `no-proxy` and confirm they are come back out of the facade with `juju model-config`
3. Test unsetting the model config changes and check the facade reports the changes again.

All other logic tests are satisfied in unit tests.

## Links

**Jira card:** JUJU-5342

### CODE REVIEW

File: apiserver/facades/client/modelconfig/backend.go [hunk 1]
```diff
@@ -18,8 +18,6 @@ type Backend interface {
 	common.BlockGetter
 	ControllerTag() names.ControllerTag
 	ModelTag() names.ModelTag
-	ModelConfigValues() (config.ConfigValues, error)
-	UpdateModelConfig(map[string]interface{}, []string, ...state.ValidateConfigFunc) error
 	Sequences() (map[string]int, error)
 	SpaceByName(string) error
```
Comment (wallyworld) [inline apiserver/facades/client/modelconfig/backend.go:22]: I wonder how much longer we need SpaceByName etc for

File: apiserver/facades/client/modelconfig/modelconfig.go [hunk 2]
```diff
@@ -157,62 +161,46 @@ func (c *ModelConfigAPI) ModelSet(ctx context.Context, args params.ModelSet) err
...
 }
 
-// checkAuthorizedKeys checks that the passed config attributes does not
-// contain authorized-keys.
-func (c *ModelConfigAPI) checkAuthorizedKeys() state.ValidateConfigFunc {
-	return func(updateAttrs map[string]interface{}, removeAttrs []string, oldConfig *config.Config) error {
-		if _, found := updateAttrs[config.AuthorizedKeysKey]; found {
-			return errors.New("authorized-keys cannot be set")
+// LogTracingValidator is a logging config validator that checks if a logging
+// change is being requested. Specifically that of trace and to see if the
```
Comment (wallyworld) [inline apiserver/facades/client/modelconfig/modelconfig.go:187]: `... requested, specifically ...`

File: apiserver/facades/client/modelconfig/service.go [hunk 3]
```diff
@@ -5,8 +5,17 @@ package modelconfig
 
 import (
 	"context"
+
+	"github.com/juju/juju/environs/config"
 )
 
+// ModelConfigService is an interface for interacting the a models underlying
```
Comment (wallyworld) [inline apiserver/facades/client/modelconfig/service.go:12]: typo plus apostrophe

File: domain/modelconfig/validators/validators_test.go [hunk 4]
```diff
@@ -72,7 +72,71 @@ func (_ *validatorsSuite) TestCharmhubURLNoChange(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 }
 
-func (_ *validatorsSuite) TestSpaceCheckerFound(c *gc.C) {
+func (*validatorsSuite) TestAgentVersionChanged(c *gc.C) {
```
Comment (wallyworld) [inline domain/modelconfig/validators/validators_test.go:75]: We seem to be missing the authorised keys validator test

File: domain/modelconfig/service/service_test.go [hunk 5]
```diff
@@ -31,6 +31,10 @@ func (f ModelDefaultsProviderFunc) ModelDefaults(
 	return f(c)
 }
 
+func (s *serviceSuite) TestSomething(c *gc.C) {
```
Comment (wallyworld) [inline domain/modelconfig/service/service_test.go:34]: ????

File: apiserver/facades/client/modelconfig/modelconfig_test.go [hunk 6]
```diff
@@ -150,89 +160,50 @@ func (s *modelconfigSuite) TestBlockChangesModelSet(c *gc.C) {
...
-	// It's okay to pass config back with the same agent-version.
-	result, err := api.ModelGet(context.Background())
-	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(result.Config["agent-version"], gc.NotNil)
-	args.Config["agent-version"] = result.Config["agent-version"].Value
-	err = api.ModelSet(context.Background(), args)
-	c.Assert(err, jc.ErrorIsNil)
-}
-
-func (s *modelconfigSuite) TestModelSetCannotChangeCharmHubURL(c *gc.C) {
```
Comment (wallyworld) [inline apiserver/facades/client/modelconfig/modelconfig_test.go:175]: These should be moved to the service as a smoke test to ensure the validators have been wired in. The individual validators can be unit tested fully, but we should have a smoke test for each one at the service level

**Review by wallyworld — APPROVED** [review]
Out of date mock to fix.
Plus I'd like to see some smoke tests to check validation wiring up at the service level.

File: apiserver/facades/client/modelconfig/service.go (hunk 3 above)
Comment (tlm) [reply apiserver/facades/client/modelconfig/service.go:12]: Done & Done

File: domain/modelconfig/service/service_test.go (hunk 5 above)
Comment (tlm) [reply domain/modelconfig/service/service_test.go:34]: Done & Done

File: domain/modelconfig/validators/validators_test.go (hunk 4 above)
Comment (tlm) [reply domain/modelconfig/validators/validators_test.go:75]: Done & Done

File: apiserver/facades/client/modelconfig/service.go (hunk 3 above)
Comment (tlm) [reply apiserver/facades/client/modelconfig/service.go:12]: Done & Done

File: apiserver/facades/client/modelconfig/modelconfig.go (hunk 2 above)
Comment (tlm) [reply apiserver/facades/client/modelconfig/modelconfig.go:187]: Done & Done

File: apiserver/facades/client/modelconfig/backend.go (hunk 1 above)
Comment (tlm) [reply apiserver/facades/client/modelconfig/backend.go:22]: Done & Done

File: domain/modelconfig/bootstrap/bootstrap_test.go [hunk 7]
```diff
@@ -38,17 +102,17 @@ func (s *bootstrapSuite) TestSetModelConfig(c *gc.C) {
 		}, nil
 	}
 
-	cfg, err := config.New(config.NoDefaults, map[string]any{
-		"name": "wallyworld",
-		"uuid": "a677bdfd-3c96-46b2-912f-38e25faceaf7",
-		"type": "sometype",
-	})
-	c.Assert(err, jc.ErrorIsNil)
+	//cfg, err := config.New(config.NoDefaults, map[string]any{
```
Comment (manadart) [inline domain/modelconfig/bootstrap/bootstrap_test.go:105]: @tlm Can we address this block please?

**Pull request merged**